src/Featurizers/Components/FrequencyEstimator.h

```diff
@@ -0,0 +1,169 @@
+    FrequencyTrainingOnlyPolicy(void);
+
+    void fit(InputType const &input);
+    FrequencyAnnotation complete_training(void);
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Types
+    // |
+    // ----------------------------------------------------------------------
+    using FrequencyType                     = std::chrono::system_clock::duration;
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    FrequencyType _frequency;
+    InputType _lastObserved;
+
+    bool _hasSeenValues;
+};
+}
+
+/////////////////////////////////////////////////////////////////////////
+///  \typedef       FrequencyEstimator
+///  \brief         A training-only class that finds the min, max
+///
+template <
+    size_t MaxNumTrainingItemsV=std::numeric_limits<size_t>::max()
+>
+using FrequencyEstimator                       = TrainingOnlyEstimatorImpl<Details::FrequencyTrainingOnlyPolicy, MaxNumTrainingItemsV>;
+
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// |
+// |  Implementation
+// |
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+
+// ----------------------------------------------------------------------
+// |
+// |  FrequencyAnnotation
+// |
+// ----------------------------------------------------------------------
+inline FrequencyAnnotation::FrequencyAnnotation(FrequencyAnnotation::FrequencyType value) :
+    Value(std::move(value)) {
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  Details::FrequencyTrainingOnlyPolicy
+// |
+// ----------------------------------------------------------------------
+Details::FrequencyTrainingOnlyPolicy::FrequencyTrainingOnlyPolicy(void) :
+    _frequency(std::chrono::system_clock::duration::max().count()),
+    _hasSeenValues(false) {
+}
+
+void Details::FrequencyTrainingOnlyPolicy::fit(InputType const &input) {
+    if (!_hasSeenValues) {
+        _lastObserved = input;
+        _hasSeenValues = true;
+        return;
+    }
+    if (_lastObserved >= input) {
+        throw std::runtime_error("Input stream not in chronological order.");
+    }
+    FrequencyType currFrequency(input - _lastObserved);
+    if (currFrequency <= _frequency) {
+        _frequency = currFrequency;
+    }
+    _lastObserved = input;
+}
+
+FrequencyAnnotation Details::FrequencyTrainingOnlyPolicy::complete_training(void) {
+    return FrequencyAnnotation(std::move(_frequency));

```

Comment:
>return FrequencyAnnotation(std::move(_frequency)); [](start = 4, length = 50)

Should it be an error to complete training without having seen a value?

Reply:
Yes it should be. Jin, I fixed this in the original branch we started working on. Can you make sure to pull in the changes/tests from that branch?